File: Day70/Part4/fog_path.py
from collections import deque

def solve():
    try:
        line = input().split()
        if not line: return
        R = int(line[0])
        C = int(line[1])
        
        grid = []
        for _ in range(R):
            grid.append(list(map(int, input().split())))
            
        line = input().split()
        startR, startC, endR, endC = map(int, line)
        
        if grid[startR][startC] == 1 or grid[endR][endC] == 1:
            print("False")
            return
            
        visited = set()
        q = deque([(startR, startC)])
        visited.add((startR, startC))
        
        dr = [0, 0, 1, -1]
        dc = [1, -1, 0, 0]
        
        while q:
            r, c = q.popleft()
            
            if r == endR and c == endC:
                print("True")
                return
            
            for i in range(4):
                nr, nc = r + dr[i], c + dc[i]
                
                if 0 <= nr < R and 0 <= nc < C:
                    if grid[nr][nc] == 0 and (nr, nc) not in visited:
                        visited.add((nr, nc))
                        q.append((nr, nc))
                        
        print("False")
    except EOFError:
        pass

if __name__ == "__main__":
    solve()

File: Day70/Part4/fog_path.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <tuple>

using namespace std;

int main() {
    int R, C;
    if (cin >> R >> C) {
        vector<vector<int>> grid(R, vector<int>(C));
        for (int i = 0; i < R; ++i)
            for (int j = 0; j < C; ++j) cin >> grid[i][j];

        int startR, startC, endR, endC;
        cin >> startR >> startC >> endR >> endC;

        if (grid[startR][startC] == 1 || grid[endR][endC] == 1) {
            cout << "False" << endl;
            return 0;
        }

        vector<vector<bool>> visited(R, vector<bool>(C, false));
        queue<pair<int, int>> q;

        q.push({startR, startC});
        visited[startR][startC] = true;

        int dr[] = {0, 0, 1, -1};
        int dc[] = {1, -1, 0, 0};

        while (!q.empty()) {
            auto [r, c] = q.front();
            q.pop();

            if (r == endR && c == endC) {
                cout << "True" << endl;
                return 0;
            }

            for (int i = 0; i < 4; ++i) {
                int nr = r + dr[i];
                int nc = c + dc[i];

                if (nr >= 0 && nr < R && nc >= 0 && nc < C) {
                    if (grid[nr][nc] == 0 && !visited[nr][nc]) {
                        visited[nr][nc] = true;
                        q.push({nr, nc});
                    }
                }
            }
        }

        cout << "False" << endl;
    }
    return 0;
}
